Popover: Refactor popover to clarify computations

youknowriad opened the pull request:
and support setting max-width/max-height depending on the available space

This will fix the issues we have where popovers are cut over the edges of the browser window by adapting the height/width of the popovers.

Still having trouble with a z-index issue at the moment.

closes #4519, #6026, #6026, #4949, #5276, #6504,

youknowriad committed:
Popover: Refactor popover to clarify computations and support setting max-width/max-height depending on the available space

youknowriad committed:
Fix popover z-index

youknowriad committed:
Adding unit tests

jorgefilipecosta commented:
Automatic test cases need an update, but manually testing things are looking good and I did not notice any problem in my initial tests :)

youknowriad committed:
Fix inserter popover

youknowriad commented on components/popover/index.js:
```diff
@@ -141,83 +133,105 @@ class Popover extends Component {
-			forcedXAxis || xAxis,
-		];
+		if (
+			! this.state.popoverLeft ||
```
Yes, it's a leftover :)

youknowriad commented on components/popover/index.js:
```diff
@@ -90,8 +79,15 @@ class Popover extends Component {
+		if ( event.type === 'scroll' && this.nodes.content.contains( event.target ) ) {
+			return;
+		}
+		this.rafHandle = window.requestAnimationFrame( () => this.computePopoverPosition() );
```
It's already the case I believe

youknowriad committed:
Remove useless check

youknowriad commented:
I fixed the tests, extracted the computing to an utility to ease unit-testing. I'm adding a design feedback. I think this improves things above popovers. It's not solving everything but it's a good improvement I think.

youknowriad committed:
Fix e2e tests

youknowriad committed:
Fix unit tests

youknowriad commented on components/popover/index.js:
```diff
@@ -31,54 +32,42 @@ const { ESCAPE } = keycodes;
+		this.computePopoverPosition( popoverSize );
 		this.toggleWindowEvents( true );
-		this.focus();
+		setTimeout( () => this.focus() );
```
I'd love to get rid of this timeout but for some mysterious reason, it's not working properly without it.

 - Try removing the timeout
 - Open the inserter
 - The search input is not focused.

I tried debugging and for some reason the call to `input.focus()` is not focusing the input (console logging the `document.actiiveElement`.

jorgefilipecosta replied:
I tried to debug, and I did not find a good solution to this, It looks like although the parent component was mounted the child is not right away in a state that we can focus, this discussion seems related https://stackoverflow.com/questions/35522220/react-ref-with-focus-doesnt-work-without-settimeout-my-example.

What complicates, even more, the things is that if we use withSafeTimeout, we create the z-index bug ( i tested it again).

Maybe we can move the setTimeout inside the focus function so inside the timeout we don't reference this, props, etc.. we reference the dom node something like:
```
		const focusNode = ( domNode ) => setTimeout( () => domNode.focus() );
		const domNode = this.contentNode.current;
		// Find first tabbable node within content and shift focus, falling
		// back to the popover panel itself.
		const firstTabbable = focus.tabbable.find( domNode )[ 0 ];
		focusNode( firstTabbable ? firstTabbable : domNode );
```

youknowriad replied:
I don't expect this to create any issue regarding "this" because it's during mouting but even though, I like your proposal let's move it to `focus` and add a comment.

jorgefilipecosta replied:
Maybe we can also add a comment specifying why we are using a set timeout, and why withSafeTimeout was not possible to use.

youknowriad commented:
I think we get this in as it solves a lot of issues. thoughts @jorgefilipecosta

youknowriad committed:
All the popovers should show up above the UI

youknowriad committed:
Leave more space for the arrow

youknowriad commented on editor/components/rich-text/index.js:
```diff
@@ -794,7 +794,7 @@ export class RichText extends Component {
 		forEach( formats, ( formatValue, format ) => {
 			if ( format === 'link' ) {
-				if ( formatValue !== undefined ) {
+				if ( !! formatValue ) {
```
The relation is indirect. There's a bug in master where if you click "Escape" on a newly created link (empty), the modal is not closed. This change ensures the modal is closed when you do so.

The reason it's in this PR is that the popover change makes the e2e test fail while it's not failing in master (false positive). I'm not certain I understand exactly why it's not failing in master tests because, in my manual testing, it's not working properly.

jorgefilipecosta replied:
Thank you for the clarification, e2e tests on our CI sometimes pass when they should fail. For example in https://github.com/WordPress/gutenberg/pull/6877 I added a test case that should fail because the fix was not yet merged, running tests in my computer the test fails all the time, but in our CI it is reported as passing, not certain why.

jorgefilipecosta commented on components/popover/test/utils.js:
```diff
@@ -0,0 +1,136 @@
+		expect( computePopoverPosition( anchorRect, contentSize, 'top right' ) ).toEqual( {
+			contentHeight: null,
+			contentWidth: null,
+			isMobile: false,
```
All the test cases test the non-mobile case should we add a mobile test case?

youknowriad replied:
Yep, good catch it was less important but I'll see if I can tweak the window size.

jorgefilipecosta commented on components/popover/utils.js:
```diff
@@ -0,0 +1,86 @@
+	};
+	const bottomAlignment = {
+		popoverTop: anchorRect.bottom,
+		contentHeight: anchorRect.bottom + HEIGHT_OFFSET + height > window.innerHeight ? window.innerHeight - HEIGHT_OFFSET - anchorRect.bottom : height,
```
Maybe window.innerHeight, and window.innerWidth can be parameters with default values to the window properties. I think it may simplify the addition of test cases in the future.

jorgefilipecosta commented on components/popover/utils.js:
```diff
@@ -0,0 +1,86 @@
+	// Choosing the x axis
+	let chosenXAxis;
+	let contentWidth = null;
+	if ( xAxis === 'center' && centerAlignment.contentWidth === width ) {
```
Minor: Maybe we can use a switch.

On the switches we would do:
```
case 'center':
	chosenXAxis = centerAlignment.contentWidth === width ? 'center' : undefined;
```

After the switch, we would check if chosenXAxis was undefined if yes we would use our last else. It is just a personal preference normally I prefer switches instead of a series of else.

It may also make sense to extract some parts of the function into smaller ones e.g: one part that deals with the x and other that deals with y, just to make computePopoverPosition smaller :)

youknowriad replied:
I don't like the switch in this case personally, I like "dumb" code :P but I like the splitting proposal which I did.

jorgefilipecosta reviewed (approved):
Excellent work on this refactor. Testing the popover it looks like a big improvement. 👍 
There is some debugging that could be made like understand why withSafeTimeout interferes with z-index if time allows in the future. But given the number of issues closed and the noticeable improvements for the user that we have I feel we should get it in soon so we can test it more before the next release.

youknowriad committed:
Set timeout moved to the focus handler

youknowriad committed:
Split the popover utility to smaller functions and test the mobile case

aduth commented on components/popover/index.js:
```diff
@@ -100,27 +96,22 @@ class Popover extends Component {
 
+		// Without the setTimeout, the dom node is not being focused
+		// Related https://stackoverflow.com/questions/35522220/react-ref-with-focus-doesnt-work-without-settimeout-my-example
+		const focusNode = ( domNode ) => setTimeout( () => domNode.focus() );
```
Where are we canceling this timeout if the component is unmounted? Why aren't we using [`withSafeTimeout`](https://github.com/WordPress/gutenberg/tree/master/components/higher-order/with-safe-timeout) ?

youknowriad replied:
I did use `withSafeTimeout` originally but It breaks the design in a super weird way. (see 66b34440947efedcedff6eb38775c82e1f01ac81) The usage of the higher order component even if it's not doing anything to the DOM was creating some weird z-index issues, especially visible on mobile.

aduth replied:
What about: 

>Where are we canceling this timeout if the component is unmounted?

youknowriad replied:
This is already done in master

aduth commented on components/popover/index.js:
```diff
@@ -100,27 +96,22 @@ class Popover extends Component {
 			return;
 		}
 
+		// Without the setTimeout, the dom node is not being focused
```
This is treating the symptom, not the cause, and the comment reads as such.

The real cause is a combination of:

- When initially mounted, the popover doesn't have a Slot context in which to render. When this becomes available, it re-renders the content into the slot (an unmount and remount operation).
- When initially mounted, the container has a style of `visibility: hidden`, so focusing would not select the container.

aduth replied:
See also: #8218

aduth commented on components/popover/index.js:
```diff
@@ -100,27 +96,22 @@ class Popover extends Component {
 
-		const { content } = this.nodes;
-		if ( ! content ) {
+		if ( ! this.contentNode.current ) {
```
Related to [earlier comment](https://github.com/WordPress/gutenberg/pull/6799/files#r206211471), we shouldn't need to check this here. If the component causes `Popover#focus` to be called during its `componentDidMount`, we should be able to assume that the `ref.current` is assigned.

aduth commented on components/popover/index.js:
```diff
@@ -90,8 +79,15 @@ class Popover extends Component {
+	}
+
+	throttledComputePopoverPosition( event ) {
+		if ( event.type === 'scroll' && this.contentNode.current.contains( event.target ) ) {
```
It took me more than a minute to grasp why this condition is here. It seems obvious in retrospect, but a comment would have been nice.
